debezium-core/src/main/java/io/debezium/transforms/ByLogicalTableRouter.java

```diff
@@ -0,0 +1,333 @@
+ * some identifier added to the key that distinguishes the different physical tables. The field name specified by
+ * the config option {@link #PHYSICAL_TABLE_IDENTIFIER_FIELD_NAME} (or {@link #DEFAULT_PHYSICAL_TABLE_IDENTIFIER_FIELD_NAME}
+ * if not specified) is added to the key schema for this purpose. By default, its value will be the old topic name, but
+ * if a custom value is desired, the config options {@link #PHYSICAL_TABLE_REGEX} and {@link #PHYSICAL_TABLE_REPLACEMENT}
+ * are used to change it. For instance, in our above example, we might choose to make the identifier `db_shard1` and
+ * `db_shard2` respectively.
+ *
+ * @author David Leibovic
+ */
+public class ByLogicalTableRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private Pattern logicalTableRegex;
+    private String logicalTableReplacement;
+    private Pattern physicalTableRegex;
+    private String physicalTableReplacement;
+    private String physicalTableIdentifierFieldName;
+    private Cache<Schema, Schema> keySchemaUpdateCache;
+    private Cache<Schema, Schema> envelopeSchemaUpdateCache;
+    private final AvroValidator schemaNameValidator = AvroValidator.create(logger);
+
+    private static final Field LOGICAL_TABLE_REGEX = Field.create("logical.table.regex")
+            .withDisplayName("Logical table regex")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withValidation(Field::isRequired, Field::isRegex)
+            .withDescription("The regex used for matching the name of the logical table. This regex will be matched " +
+                "against the original topic name.");
+
+    private static final Field LOGICAL_TABLE_REPLACEMENT = Field.create("logical.table.replacement")
+            .withDisplayName("Logical table replacement")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withValidation(Field::isRequired)
+            .withDescription("The replacement string used in conjunction with " + LOGICAL_TABLE_REGEX.name() +
+                    ". This will be used to create the new topic name.");
+
+    private static final Field PHYSICAL_TABLE_REGEX = Field.create("physical.table.regex")
+            .withDisplayName("Physical table regex")
+            .withType(ConfigDef.Type.STRING)
+            .withWidth(ConfigDef.Width.LONG)
+            .withImportance(ConfigDef.Importance.LOW)
+            .withValidation(Field::isRegex)
+            .withDescription("Now that multiple physical tables can share a topic, the Key Schema can no longer " +
+                "consist of solely the record's primary / unique key fields, since they are not guaranteed to be " +
```

Comment:
It may be _likely_ that one has to augment the key, but strictly speaking augmenting the key is not required in all cases. Perhaps `...share a topic, the event's key may need to be amended to include fields other than just those for the record's primary/unique, since these are not guaranteed ...`